feat(android): locale fixes and improvements

prashantsaini1 opened the pull request:
This PR brings implement following changes for Android platform:

1. Brings following stability in `Ti.Locale.setLanguage()` method:
    - Fixes a crash where apps, with multiple windows opened, cannot be restarted properly with `Ti SDK 13.0.0.GA` and `targetSdkVersion: 35`.
    - When changing the language using this method, the SDK will restart the apps automatically unless if a `Ti.Locale.change` event is added.
    - **The SDK will persist this setting across app restarts.**
2. Handles the system language updates properly on all versions, especially Android 13+.
3. Parity with iOS in `Ti.Locale.current*` APIs to return the app's current data. Earlier, these APIs reports system only data on Android 13+ until the `Ti.Locale.setLanguage` is called on every app launch by JS side.

_Tested on Android 10, Android 13 and Android 15 apps._

Some error logs from a crash:
```
16:11:45.599 Proxy           Proxy.prototype.getProperty is not a Function!
16:11:45.599 Proxy           Proxy.prototype.setProperty isn't a function!!!
16:11:45.599 Proxy           Proxy.prototype.getProperty is not a Function!
16:11:45.599 Proxy           Proxy.prototype.getProperty is not a Function!
16:11:45.601 TiExce...andler (main) [6,6] /alloy/controllers/BaseController.js:436
                             throw new Error(`Passed proxy "${proxy.id}" is not a valid Alloy proxy. Attempted to add event listener for type = "${type}"`);
                             ^
                             Error: Passed proxy "undefined" is not a valid Alloy proxy. Attempted to add event listener for type = "click"
                                 at Controller.addListener (/alloy/controllers/BaseController.js:436:7)
                                 org.appcelerator.kroll.runtime.v8.V8Function.nativeInvoke(Native Method)
                                 org.appcelerator.kroll.runtime.v8.V8Function.callSync(V8Function.java:55)
                                 org.appcelerator.kroll.runtime.v8.V8Function.call(V8Function.java:41)
16:11:45.601 V8Exception     Exception occurred at /alloy/controllers/BaseController.js:436: Uncaught Error: Passed proxy "undefined" is not a valid Alloy proxy. Attempted to add event listener for type = "click"
16:11:46.355 Proxy           Proxy.prototype.getProperty is not a Function!
16:11:46.355 Proxy           Proxy.prototype.setProperty isn't a function!!!
16:11:46.355 Proxy           Proxy.prototype.getProperty is not a Function!
16:11:46.355 Proxy           Proxy.prototype.getProperty is not a Function!
```

prashantsaini1 committed:
feat(android): locale crash fixes, and parity with iOS

hansemannn commented on apidoc/Titanium/Locale/Locale.yml:
```diff
@@ -235,3 +251,45 @@ properties:
     type: String
     permission: read-only
+
+  - name: systemCountry
```
This should be unified via `Ti.Locale.currentCountry`. If something is overridden, it should be reflected in the current country.

hansemannn commented on apidoc/Titanium/Locale/Locale.yml:
```diff
@@ -235,3 +251,45 @@ properties:
+    permission: read-only
+    since: "13.1.0"
+
+  - name: systemLanguage
```
This should be unified via `Ti.Locale.currentLanguage`. If something is overridden, it should be reflected in the current language.

hansemannn commented on apidoc/Titanium/Locale/Locale.yml:
```diff
@@ -235,3 +251,45 @@ properties:
+    permission: read-only
+    since: "13.1.0"
+
+  - name: systemLocale
```
Same as above.

hansemannn reviewed (changes requested):
We have to reuse existing APIs, not add more platform-specific ones.

prashantsaini1 commented:
@hansemannn Android 13+ manages system specific and app specific locales/languages separately.

- The existing `Ti.Locale.current*` APIs only returns the app specific data on all Android versions.
    - e.g. If system is using `en`, and an app is using `de`, then `Ti.Locale.currentLanguage` returns `de`.
    - There're no APIs to get the system specific data for Android 13+.

If we merge this behaviour in `Ti.Locale.current*` APIs, then users won't be able to know app's current language on Android 13+ and will break the current expected behaviour.

hansemannn commented:
> Android 13+ manages system specific and app specific locales/languages separately.

iOS does the same since iOS 15/16, but the system handles it correctly there -> If the system language is overridden, that value is returned. The same should happen for Android.

prashantsaini1 commented:
> iOS does the same since iOS 15/16, but the system handles it correctly there -> If the system language is overridden, that value is returned. The same should happen for Android.

`Ti.Locale.currentLanguage` does not return the overridden value on Android 13+. This PR fixes this exact issue and brings parity with iOS, along with the additional handling of auto restarting the apps at right time.

- I will keep this PR for crash and behaviour fixes only.
- An additional PR for new props/methods. Seems for iOS also we do not have any existing API to know the system language.

prashantsaini1 committed:
chore: format docs

m1ga commented:
sadly I can't test this. On my phone it's not doing any restart when changing the language, even with 13.0.0.GA and the `change` event it also not triggered. Have to check why it's like this before I can run some tests here.

One thing you should also verify: if you set the per-app language in the app settings it should use that when you start your app

prashantsaini1 commented:
> On my phone it's not doing any restart when changing the language, even with 13.0.0.GA and the change event it also not triggered. Have to check why it's like this before I can run some tests here.

- Probably the locale broadcast receiver (in TiApplication) is not getting registered, or triggered at least, likely due to Battery Optimizations. Android 13+ relies on broadcast to trigger app restart or the `change` event.
- Moving the `TiLocaleManager.handleSystemLocaleUpdates();` to TiApplication's `onConfigurationChanged` should work fine on your device. Can you please verify it?
- If the above works, then the code needs more handling since `onConfigurationChanged` can be called due to various reasons and we may not want to always restart the apps.

> One thing you should also verify: if you set the per-app language in the app settings it should use that when you start your app

I think it should be fine because the PR will keep using the system defined language or the per-app language until it's overriden by user by calling the `Ti.Locale.setLanguage()`.

hansemannn commented:
I would highly prefer not to merge this with the additional properties. Again, iOS also manages it in one property and there is no practical use case so far to mainatin both under own, platform specific / Android only properties.

prashantsaini1 commented:
@hansemannn I've already removed additional properties from this PR as we only need the `setLanguage` crash fix for now.

hansemannn commented:
Ah, thank you! Once Michael also approves it, it's fine for me then.

prashantsaini1 commented on android/titanium/src/java/org/appcelerator/titanium/util/TiLocaleManager.java:
```diff
@@ -6,34 +6,162 @@
+	{
+		try {
+			return TiApplication.getInstance().getAppProperties().getString(KEY_LOCALE, null);
+		} catch (Exception ignoredException) {
```
We really want to ignore this exception as we handle it further in another way.

prashantsaini1 commented on android/titanium/src/java/org/appcelerator/titanium/util/TiLocaleManager.java:
```diff
@@ -6,34 +6,162 @@
+
+				return Locale.getDefault();
+			}
+		} catch (Exception ignoredException) {
```
We really want to ignore this exception as we handle it further in another way.

prashantsaini1 committed:
chore: handle locale changes in onConfigurationChanged rather than the broadcast

m1ga commented:
I've tested it on a Android 16 device and most of it works now. The only thing that doesn't work without a force restart now is using the per-app language setting, that works with 13.0.0.GA. So if I switch the language in the per-app setting and go back into the app it won't update the language now. Using the test code from before and the button inside the app works and it will remember it after the restart. Also no double closing now.

prashantsaini1 committed:
chore: handle per-app locale setting outside the app

prashantsaini1 commented:
@m1ga It should work now for the per-app locale as well from the Device Settings. Thanks for your prompt tests!

prashantsaini1 committed:
revert: revert to using locale broadcast, with updates handling on main-thread

`onConfigurationChanged` is called multiple times, causing app restarts in unexpected cases

prashantsaini1 commented:
@m1ga As we discussed on Slack, we cannot safely rely on onConfigurationChanged method and has to use locale broadcast. Could you please test the latest commit on your device to find out if it's not getting registered or not getting called after registration? Thanks again for your responses!

prashantsaini1 commented:
@m1ga Thanks for the updates. So it was not the broadcast registration issue, but rather performing locale updates on main thread. Glad that it works finally.

I tested the in app language switching, it already works fine at first place.  Let's roll it out!

prashantsaini1 committed:
Merge remote-tracking branch 'upstream/master' into android-per-app-locale
